Feature: Asynchronous Cleanup of Scratch Directory

reneme opened the pull request:
This Pull Request is requesting for comments!

After `cvmfs_server publish`-ing a large change set it might take a significant time to clean-out the scratch directory (i.e. `/var/spool/cvmfs/<fqrn>/scratch`). This allows to run this clean-up asynchronously to allow for quicker publishing round-trips. Setting `CVMFS_ASYNC_SCRATCH_CLEANUP=false` disables this and reverts back to the legacy behaviour.

The general idea of this workflow is:

0. create `current` and `wastebin` inside `/var/spool/cvmfs/<fqrn>/scratch`
1. publish as usual (using `.../scratch/current` as union fs scratch directory)
2. umount `/cvmfs/<fqrn>` and `/var/spool/cvmfs/<fqrn>/rdonly`
3. move `.../scratch/current` into `.../wastebin/waste.XXXXXX/current`
4. create a fresh `.../scratch/current`
5. spawn daemon process asynchronously deleting everything in `.../wastebin`
6. mount `/var/spool/cvmfs/<fqrn>/rdonly` and `/cvmfs/<fqrn>`

Note that there might be multiple concurrent `cvmfs_suid_helper clear_scratch_async` but they should not step on each other's feet. Subject to review: Is this actually true?

Theoretically back-to-back transactions should become quicker but this effect might be mitigated due to the background process stressing the file system with the recursive delete.

Furthermore this requires a repository migration step *2.2.x --> 2.3.0* to be developed.

reneme commented on cvmfs/cvmfs_suid_helper.cc:
```diff
@@ -104,64 +105,127 @@ static void KillCvmfs(const string &fqrn) {
+
+  bool success = true;
+  ScopedWorkingDirectory::DirectoryEntry dirent;
+  while (success && swd.NextDirectoryEntry(&dirent)) {
```
This stops the recursive delete as soon as there was an error, which might result in an unfinished cleanup. Whatever is left-over should be taken care of by the following publish. Not sure if we should have more safe guards here?

DrDaveD commented:
It's a good idea in general.  I do something similar in frontier-squid to asynchronously clear the cache.   I do it all in bash however, which seems a lot simpler than the rather extensive changes to run_suid_helper.  I can give you the details if you're interested.

I don't see anything about catching signals.  In frontier-squid I run the rm with nohup so they'll keep going after a user logs out.  I think that's important.

I suggest reversing the sense of the variable to be CVMFS_SYNC_SCRATCH_CLEANUP so it can default to false and not require a migration step.  I doubt that anybody will have a problem with the default and want to change it.

reneme commented:
> I do it all in bash however, which seems a lot simpler than the rather extensive changes to run_suid_helper.

Unfortunately we need to do it in `cvmfs_suid_helper` because the scratch area might contain files with skewed permissions (f.e. hidden housekeeping stuff from the union file system). The changes to this code aren't so extensive, its mostly a refactoring of the recursive delete.

> I don't see anything about catching signals.

Frankly I'm not sure, but do I need to do that after a double-fork?

> I suggest reversing the sense of the variable to be CVMFS_SYNC_SCRATCH_CLEANUP so it can default to false and not require a migration step.

That's a good idea. I also thought about making this more transparent. That will require a bit of extra code in `cvmfs_server` but perhaps it is more reasonable than the explicit migration step.

Thanks for the comments.

DrDaveD commented:
> Unfortunately we need to do it in cvmfs_suid_helper because the scratch area might contain files with skewed permissions (f.e. hidden housekeeping stuff from the union file system). The changes to this code aren't so extensive, its mostly a refactoring of the recursive delete.

I wasn't suggesting to not use cvmfs_suid_helper, but it could be cvmfs_suid_helper &

> Frankly I'm not sure, but do I need to do that after a double-fork?

I'm pretty sure you do, unless you call setsid() to give it a separate process group.  Try it.

reneme commented:
> I'm pretty sure you do, unless you call setsid() to give it a separate process group. Try it.

It seems to be fine (Ubuntu 15.10) with just this double-fork but anyway I've added `setsid()`.

Furthermore I've added a `cvmfs_server migrate` step for the change in `/var/spool/cvmfs/<repo>/scratch`. We won't easily get around that, I'm afraid.

jblomer commented on cvmfs/cvmfs_server:
```diff
@@ -5624,6 +5650,47 @@ migrate_2_1_20() {
+  echo "--> umount repository"
+  run_suid_helper rw_umount $name
+
+  echo "--> updating scratch directory layout"
```
Can this break if the server package is update while a transaction is open?

reneme replied:
Potentially... We could check for open transactions in `migrate()` but that is going to cause a dead-lock. `cvmfs_server` will prevent you from closing the transaction with `cvmfs_server abort` if the version number doesn't fit.

Anyway: That might be more favourable than deleting the scratch directory which might contain valuable data.

reneme replied:
I'll submit a patch for that in another pull request.

jblomer commented on cvmfs/cvmfs_server:
```diff
@@ -5624,6 +5650,47 @@ migrate_2_1_20() {
+  echo "--> updating /etc/fstab"
+  local tmp_fstab=$(mktemp)
+  local comment="added by CernVM-FS for ${name}"
+  sed -e "s~^\(.*\)\(${scratch_dir}\)\(.*${comment}\)$~\1\2/current\3~" /etc/fstab > $tmp_fstab
```
Can we use `sed -i -e` and avoid handling the temp file?

reneme replied:
[mais, bien sûre](https://github.com/cvmfs/cvmfs/pull/1541/commits/f29e4b2f83cde42921a42709cbb8c02d150db98e).

jblomer commented on cvmfs/cvmfs_suid_helper.cc:
```diff
@@ -104,64 +105,129 @@ static void KillCvmfs(const string &fqrn) {
+    assert(retval == 0);
+  }
+
+  bool IsMetaEntry(const platform_dirent64 *dirent) {
```
Perhaps `IsDotEntry`?

reneme replied:
[aye](https://github.com/cvmfs/cvmfs/pull/1541/commits/3d4610df4aa7da545f05f2eecfdfee1c5dd61d1a).

jblomer commented on cvmfs/cvmfs_suid_helper.cc:
```diff
@@ -104,64 +105,129 @@ static void KillCvmfs(const string &fqrn) {
+    int retval = setsid();
+    assert(retval != -1);
+    if ((pid = fork()) == 0) {
+      close(0); close(1); close(2);
```
Inside `CleanupDirectory`, we would still print to `stderr` if things fail, no?

reneme replied:
Mhh, wouldn't that pollute a random shell that potentially is long gone? My hope would be that consecutive runs of this cleanup would free such things. However, some kind of log would probably make sense. Syslog?

reneme replied:
[Fixed](https://github.com/cvmfs/cvmfs/pull/1541/commits/4763580be4b970d575b110a6168affc77445c1a0) by code duplication after all. I didn't factor out `Daemonize()` in `util/posix.h` because it calls to `LogCvmfs()` whose dependency I'd like to avoid.
